batteries/difftext/myersdiff.luau

```diff
@@ -0,0 +1,132 @@
+local deque = require("@batteries/collections/deque")
+local types = require("./types")
+
+type diffoperation = types.diffoperation
+type diff = types.diff
+
+type EditGraphNode = {
+	x: number,
+	y: number,
+	prev: EditGraphNode?,
+}
+
+local function editGraphNode(x: number, y: number, prev: EditGraphNode?): EditGraphNode
+	return {
+		x = x,
+		y = y,
+		prev = prev,
+	}
+end
+
+-- true if a[x+1] == b[y+1]; indicates an EQUAL operation (no diff) is valid
+local function hasDiagonalEdge(node: EditGraphNode, a: string | { string }, b: string | { string }): boolean
+	local x, y = node.x, node.y
+	if typeof(a) == "table" and typeof(b) == "table" then
+		return x < #a and y < #b and a[x + 1] == b[y + 1]
+	end
+	a = a :: string
+	b = b :: string
+	return x < #a and y < #b and a:sub(x + 1, x + 1) == b:sub(y + 1, y + 1)
+end
+
+local function edgeTodiff(curNode: EditGraphNode, a: string | { string }, b: string | { string }): diffoperation?
+	-- maps stepping from curNode.prev -> curNode to a diff operation
+	-- (n, n) -> (n+1, n) == DELETION of a[n]
+	-- (n, n) -> (n, n+1) == ADDITION of b[n]
+	-- (n, n) -> (n+1, n+1) == EQUAL (no diff)
+	if not curNode.prev then
+		return nil
+	end
+
+	local prev = curNode.prev
+	local xdiff = curNode.x - prev.x
+	local ydiff = curNode.y - prev.y
+	if xdiff == 1 and ydiff == 1 then
+		return {
+			key = "EQUAL",
+			text = if typeof(a) == "table" then a[curNode.x] else a:sub(curNode.x, curNode.x),
+		}
+	elseif xdiff == 1 then
+		return {
+			key = "DELETE",
+			text = if typeof(a) == "table" then a[curNode.x] else a:sub(curNode.x, curNode.x),
+		}
+	else
+		return {
+			key = "ADD",
+			text = if typeof(b) == "table" then b[curNode.y] else b:sub(curNode.y, curNode.y),
+		}
+	end
+end
+
+local function myersdiff(a: string | { string }, b: string | { string }): diff
+	local start = editGraphNode(0, 0)
+	local editGraphDeque = deque.new(start)
+
+	local diff = {} :: diff
+	local seen = {} :: { [string]: true }
+	seen["0,0"] = true
+
+	while #editGraphDeque do
+		local curNode = editGraphDeque:popfront()
+
+		-- we've reached bottom right vertex; indicates full edit from a -> b
+		if curNode.x == #a and curNode.y == #b then
+			-- reconstruct diff operations by tracing path from termination node to start
+			while curNode.prev do
+				local diffOperation = edgeTodiff(curNode, a, b) :: diffoperation
+				table.insert(diff, diffOperation)
+				curNode = curNode.prev
+			end
```

Comment:
let's try not to use typecasts if we can; they clobber the information the type system is reporting
we could instead do something like
```luau
local currOp = edgeToDiff(curNode, a, b)
while currOp do
    table.insert(diff, currOp)
    curNode = curNode.prev
    currOp = edgeToDiff(curNode, a, b)
end
```